src/ir/auth_logic/ast_construction.cc

```diff
@@ -128,6 +128,40 @@ static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
                        *CHECK_NOTNULL(can_say_fact_context.fact()));
 }
 
+static ArgumentType ConstructArgumentType(
+    AuthLogicParser::AuthLogicTypeContext& auth_logic_type_context) {
+  if (auto* number_type_context =
+          dynamic_cast<AuthLogicParser::NumberTypeContext*>(
+              &auth_logic_type_context)) {
+    return ArgumentType(ArgumentType::Kind::kNumber);
+  } else if (auto* principal_type_context =
+                 dynamic_cast<AuthLogicParser::PrincipalTypeContext*>(
+                     &auth_logic_type_context)) {
+    return ArgumentType(ArgumentType::Kind::kPrincipal);
+  }
+  auto& custom_type_context =
+      *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::CustomTypeContext*>(
+          &auth_logic_type_context));
+  return ArgumentType(ArgumentType::Kind::kCustom,
+                      custom_type_context.ID()->getText());
+}
+
+static RelationDeclaration ConstructRelationDeclaration(
+    AuthLogicParser::RelationDeclarationContext& relation_declaration_context) {
+  std::vector<std::pair<std::string, ArgumentType>> arguments;
```

Comment:
If you know the size of a `vector` in advance (as you do here), it's a cheap and safe optimization to call `reserve` with the size that you expect the vector to be.